pandas/core/indexes/base.py

```diff
@@ -2228,27 +2264,37 @@ def union(self, other):
             other = other.astype('O')
             return this.union(other)
 
+        # TODO: setops-refactor, clean all this up
```

Comment:
this if/then needs to all be completely removed

Reply:
Can't be removed until setops are fix more generally. [Like I said](https://github.com/pandas-dev/pandas/pull/19558#issuecomment-364280424), I have a PR started, but it's a bigger issue than what's on the critical path for ExtensionArray.

Take `CategoricalIndex.intersection` We *relied* on `CategoricalIndex._outer_indexer(self._values, other._values)` raising a TypeError to get the correct result. If I pass `CategoricalIndex._ndarray_values`, we don't get a type error any more, but we get the wrong result.

There's no way to avoid the if statements until set ops are completely refactored.

Reply:
ok, let's do this next then. these if/then/else are really smelly.

Reply:
if *might* be worth it to put this in a single method now (not sure if its possible) to isolate this code, maybe

```
def _get_internal_type(other):
     if is_period_dtype(other) ......
          ...
    else:
         ...
```